### ISSUE TITLE
Investigate improvements to NLU performance

### ISSUE THREAD

**Issue by akelad** [RasaHQ/rasa-demo#460]
Since Rasa 1.7.0 is released, we should upgrade to it (but not upload the model to Rasa X yet). 
We should also experiment with how much we can improve NLU performance, because some performance issues were fixed with 1.7.0

Specifically try taking out the loss_type here and see how the histograms look like:
- name: EmbeddingIntentClassifier
  epochs: 50
  loss_type: margin

Also good to try out how ConveRT does etc

**Comment by erohmensing** [comment]
Update is being made here https://github.com/RasaHQ/rasa-demo/issues/460

But experimenting with NLU performance after being back on softmax and also checking out convert is definitely still worthwhile!

**Comment by melindaloubser1** [comment]
comparison of current pipeline & convert pipelines the shows convert has best performance. Ignore "config_noloss" -should be "config_marginloss".
 
[nlu_model_comparison_graph.pdf](https://github.com/RasaHQ/rasa-demo/files/4182376/nlu_model_comparison_graph.pdf)

**Comment by akelad** [comment]
@melindaloubser1 can you also post the histograms of each of these? I'm curious to see what kind of confidences we have between wrong and correct classifications

**Comment by melindaloubser1** [comment]
These are all from 5-fold cv since the comparison doesn't output histograms (or at least it didn't the way I ran it?)


## config
![hist](https://user-images.githubusercontent.com/32034278/74252663-f7e81980-4cbb-11ea-8b58-a5bf6c3ad4cf.png)

## config_marginloss
![hist](https://user-images.githubusercontent.com/32034278/74252632-f1f23880-4cbb-11ea-925a-a961c36ca3db.png)

## config_convert

![hist](https://user-images.githubusercontent.com/32034278/74252591-e43cb300-4cbb-11ea-848e-edab5218e4cf.png)

**Comment by erohmensing** [comment]
Nice. Maybe we should also investigate specifically the intent prediction errors that have such high confidence?

**Comment by akelad** [comment]
yes we should! and also can you post the configs you used for each?

**Comment by melindaloubser1** [comment]
Configs:
config.yml
```yml
language: en
pipeline:
- name: WhitespaceTokenizer
- name: CRFEntityExtractor
- name: CountVectorsFeaturizer
  OOV_token: oov
  token_pattern: (?u)\b\w+\b
- name: CountVectorsFeaturizer
  analyzer: char_wb
  min_ngram: 1
  max_ngram: 4
- name: EmbeddingIntentClassifier
  epochs: 50
  ranking_length: 5
```

config_marginloss
```yml
language: en
pipeline:
- name: WhitespaceTokenizer
- name: CRFEntityExtractor
- name: CountVectorsFeaturizer
  OOV_token: oov
  token_pattern: (?u)\b\w+\b
- name: CountVectorsFeaturizer
  analyzer: char_wb
  min_ngram: 1
  max_ngram: 4
- name: EmbeddingIntentClassifier
  epochs: 50
  ranking_length: 5
  loss_type: margin
- name: DucklingHTTPExtractor
  url: http://localhost:8000
  dimensions:
  - email
  - number
  - amount-of-money
- name: EntitySynonymMapper
- name: ResponseSelector
  retrieval_intent: out_of_scope
```

config_convert
```
language: en
pipeline:
- name: ConveRTTokenizer
- name: CRFEntityExtractor
- name: ConveRTFeaturizer
- name: EmbeddingIntentClassifier
  epochs: 50
  ranking_length: 5

```

**Comment by melindaloubser1** [comment]
on cross val with `config_convert`, these are the intents where errors with confidence > 0.8 were made:
```json
{
  "affirm": 2,
  "ask_builder": 3,
  "ask_faq_languages": 2,
  "ask_faq_tutorials": 1,
  "ask_howold": 1,
  "ask_weather": 2,
  "ask_whatisrasa": 1,
  "deny": 7,
  "enter_data": 9,
  "greet": 9,
  "out_of_scope": 20,
  "react_positive": 2,
  "technical_question": 5,
  "ask_howdoing": 3,
  "contact_sales": 1,
  "how_to_get_started": 2,
  "install_rasa": 2,
  "signup_newsletter": 1,
  "switch": 2,
  "ask_faq_rasax": 2,
  "ask_faq_channels": 2,
  "ask_faq_ee": 1,
  "bye": 1,
  "human_handoff": 2,
  "nlu_info": 2,
  "ask_whoisit": 2,
  "canthelp": 1,
  "thank": 2,
  "ask_how_contribute": 1,
  "handleinsult": 1,
  "react_negative": 1
}
```

**Comment by melindaloubser1** [comment]
details for erroneous predictions with `confidence > 0.8`:
```json
{
  "affirm": {
    "errors": 2,
    "confused_with": {
      "how_to_get_started": 1,
      "enter_data": 1
    }
  },
  "ask_builder": {
    "errors": 3,
    "confused_with": {
      "out_of_scope": 2,
      "ask_whoisit": 1
    }
  },
  "ask_faq_languages": {
    "errors": 2,
    "confused_with": {
      "ask_languagesbot": 1,
      "enter_data": 1
    }
  },
  "ask_faq_tutorials": {
    "errors": 1,
    "confused_with": {
      "out_of_scope": 1
    }
  },
  "ask_howold": {
    "errors": 1,
    "confused_with": {
      "out_of_scope": 1
    }
  },
  "ask_weather": {
    "errors": 2,
    "confused_with": {
      "out_of_scope": 2
    }
  },
  "ask_whatisrasa": {
    "errors": 1,
    "confused_with": {
      "how_to_get_started": 1
    }
  },
  "deny": {
    "errors": 7,
    "confused_with": {
      "out_of_scope": 1,
      "enter_data": 3,
      "react_negative": 1,
      "install_rasa": 1,
      "handleinsult": 1
    }
  },
  "enter_data": {
    "errors": 9,
    "confused_with": {
      "out_of_scope": 5,
      "affirm": 1,
      "deny": 2,
      "how_to_get_started": 1
    }
  },
  "greet": {
    "errors": 9,
    "confused_with": {
      "out_of_scope": 4,
      "enter_data": 3,
      "affirm": 1,
      "ask_howdoing": 1
    }
  },
  "out_of_scope": {
    "errors": 20,
    "confused_with": {
      "affirm": 1,
      "enter_data": 12,
      "ask_weather": 1,
      "signup_newsletter": 2,
      "install_rasa": 1,
      "greet": 1,
      "contact_sales": 1,
      "ask_howdoing": 1
    }
  },
  "react_positive": {
    "errors": 2,
    "confused_with": {
      "affirm": 2
    }
  },
  "technical_question": {
    "errors": 5,
    "confused_with": {
      "ask_faq_is_programming_required": 1,
      "out_of_scope": 2,
      "source_code": 1,
      "how_to_get_started": 1
    }
  },
  "ask_howdoing": {
    "errors": 3,
    "confused_with": {
      "out_of_scope": 1,
      "greet": 2
    }
  },
  "contact_sales": {
    "errors": 1,
    "confused_with": {
      "signup_newsletter": 1
    }
  },
  "how_to_get_started": {
    "errors": 2,
    "confused_with": {
      "ask_whatisrasa": 1,
      "install_rasa": 1
    }
  },
  "install_rasa": {
    "errors": 2,
    "confused_with": {
      "technical_question": 1,
      "how_to_get_started": 1
    }
  },
  "signup_newsletter": {
    "errors": 1,
    "confused_with": {
      "enter_data": 1
    }
  },
  "switch": {
    "errors": 2,
    "confused_with": {
      "enter_data": 1,
      "technical_question": 1
    }
  },
  "ask_faq_rasax": {
    "errors": 2,
    "confused_with": {
      "enter_data": 1,
      "ask_faq_differencerasarasax": 1
    }
  },
  "ask_faq_channels": {
    "errors": 2,
    "confused_with": {
      "technical_question": 2
    }
  },
  "ask_faq_ee": {
    "errors": 1,
    "confused_with": {
      "out_of_scope": 1
    }
  },
  "bye": {
    "errors": 1,
    "confused_with": {
      "greet": 1
    }
  },
  "human_handoff": {
    "errors": 2,
    "confused_with": {
      "enter_data": 2
    }
  },
  "nlu_info": {
    "errors": 2,
    "confused_with": {
      "how_to_get_started": 2
    }
  },
  "ask_whoisit": {
    "errors": 2,
    "confused_with": {
      "ask_builder": 1,
      "out_of_scope": 1
    }
  },
  "canthelp": {
    "errors": 1,
    "confused_with": {
      "out_of_scope": 1
    }
  },
  "thank": {
    "errors": 2,
    "confused_with": {
      "affirm": 1,
      "ask_question_in_forum": 1
    }
  },
  "ask_how_contribute": {
    "errors": 1,
    "confused_with": {
      "ask_whatspossible": 1
    }
  },
  "handleinsult": {
    "errors": 1,
    "confused_with": {
      "out_of_scope": 1
    }
  },
  "react_negative": {
    "errors": 1,
    "confused_with": {
      "out_of_scope": 1
    }
  }
}
```

**Comment by akelad** [comment]
ok i would say let's change the default config to Convert, but would also be great @melindaloubser1 if you look into the specific examples that get confused and see if it's actually errors in our training data

**Comment by melindaloubser1** [comment]
Ok, so some results for 1.8 configs:
![image](https://user-images.githubusercontent.com/32034278/77194325-2248a600-6ab6-11ea-949e-0686a0b117d1.png)

Same configs as above + configs using DIET and a variable number of epochs. 
[configs.zip](https://github.com/RasaHQ/rasa-demo/files/4361268/configs.zip)

**Comment by melindaloubser1** [comment]
config_convert_diet_300 (on a 25% test set)
![hist](https://user-images.githubusercontent.com/32034278/77194550-9125ff00-6ab6-11ea-978c-e36707fb632c.png)

**Comment by melindaloubser1** [comment]
based on this 200 epochs seems sufficient. Since my computer was burning up trying to run this I haven't tried to run another comparison for other settings (will do in vm next time) but it looks like at least it's better than what we had before now.

**Comment by akelad** [comment]
awesome work!
